### DISCUSSION TITLE
Regarding the issue with `ubcpdk` running in kLayout

### DISCUSSION THREAD

**Discussion by Corafan** [gdsfactory/gdsfactory#3441] [Q&A] [answered]
In macro development, the `ubcpdk` file (`klayout_Library`) cannot run properly. The issue seems to be that the modules `gdsfactory` and `ubcpdk` cannot be imported. Strangely, other files in the same folder can be imported successfully. Despite trying many methods and confirming that the path is correct, errors always occur.I tried importing different modules from the same folder to identify the cause and found that gdsfactory consistently fails to import. The attempts and results are as follows:
Try to import gdsfactory:
![image](https://github.com/user-attachments/assets/9f2a8e9e-04cc-4824-a48d-a331219f83cd)
![image](https://github.com/user-attachments/assets/a7377514-9663-4a94-b916-6db605028250)
![a3308073131e19b0bb1452fedd367837](https://github.com/user-attachments/assets/0579d1d9-0303-48c4-af72-66113a278dea)
Result:
![7f1b34838f667be130fbaad568124294](https://github.com/user-attachments/assets/283beead-c37e-4ea9-b47d-971c607d8a17)
However, the file does exist:
![b31add72b9d4cf11f6f25b40e97adf3b](https://github.com/user-attachments/assets/fe49e35f-1fde-4dc2-9c2a-50e362ab76d5)
The result of importing other files (the folder containing this file is the same as the folder containing gdsfactory). And the comparison:
![9dc9fbd22bd330286be1e18f8720d834](https://github.com/user-attachments/assets/25212c44-59d9-4775-992f-3495c9d9ddf6)
![8de7e9e0be4026479ba3169aa436a762](https://github.com/user-attachments/assets/263e532d-5078-4a80-8c92-00c099e5f827)

**Comment by joamatab** [comment, marked as answer]
Hi,

GDSFactory is not designed to run directly within KLayout's macro development environment.

Instead, it is intended to be used from tools like VSCode or Jupyter Notebooks, with the GDSFactory KLayout macro serving as a bridge to communicate with KLayout.

Let me know if you have any questions or need further clarification!

**Reply by Corafan** [reply]
Hi! 

Do you mean that klayout macro development cannot import gdsfactory? Then how can UBCPDK be called in klayout macro development? If it cannot be called, does that mean ubcpdk cannot be used directly in klayout?  I have not seen any instructions for using it in the ubcpdk library.So, I want to know what is its specific usage and development method? 

Looking forward to your reply！

**Reply by joamatab** [reply]
Thank you for your question.

While it is technically possible to run Gdsfactory inside KLayout, we generally don’t recommend this approach as it can lead to complications. Instead, we suggest using Gdsfactory as a standalone tool for design automation and leveraging its integration with libraries like UBCPDK.

You can generate GDS files with Gdsfactory and then view or edit them in KLayout if needed. For guidance, I recommend referring to the Gdsfactory documentation, which provides detailed instructions on recommended usage and workflows.

Let me know if you have any other questions or need further clarification!

Best regards,

**Reply by sebastian-goeldi** [reply]
Also, as this is clearly a windows environment, it's extremely annoying installing dependencies in windows which aren't pure python packages. They need to be packages as MSYS2 packages which need to be installed more or less individually.

I highly recommend going the route Joaquin suggested. Have a separate python environment from klayout and its macro development and generate the layout there. There are utilities in the klayout gdsfactory package for seamless integration.

**Reply by Corafan** [reply]
![image](https://github.com/user-attachments/assets/2a34b3a3-eb84-44fa-8ec6-abde275a9cc7)
I would like to know how this function is implemented? I am currently unable to use this GUI driven layout function, and this video does not explain in detail how to use it. Thank you for your detailed answer

**Reply by joamatab** [reply]
That video was created as a proof of concept by a single developer, and I found the setup process to be quite challenging.

GDSFactory is flexible and capable of supporting your needs, but please note that we can't provide free support for features that are not officially documented.

If you'd like to explore custom implementations, feel free to reach out for potential consulting options or community guidance.

**Reply by Corafan** [reply]
OK, Thank you! Best wishes!
